storage/core/src/main/java/com/hortonworks/registries/storage/TransactionManager.java

```diff
@@ -27,13 +28,87 @@
+     *
+     * Implementation of this method must guarantee transaction is closed after calling this method,
+     * whether it throws Exception or not, because without guarantee, caller should try calling
+     * this method infinitely unless it succeeds, to avoid transaction left in open and possibly be leaked.
+     *
+     * With this guarantee, caller can just call a method once (and optionally handles Exception)
+     * without worrying about the status of transaction.
      */
     void rollbackTransaction();
 
 
     /**
-     * Flushes the changes made to the storage layer
+     * Flushes the changes made to the storage layer.
+     *
+     * Unlike of rollbackTransaction(), implementation of this method doesn't need to guarantee
+     * transaction is closed, since in normal caller would want to call rollbackTransaction()
+     * when commitTransaction() throws Exception.
      */
     void commitTransaction();
+
+    // TODO: javadoc
+    default <T, R> R runInTransaction(TransactionIsolation transactionIsolation,
+                                      TransactionManagerAwareFunctionCapableOfThrowingException<T, R> fn, T t) throws Exception {
+        final TransactionManager tm = this;
+        return runInTransactionInternal(transactionIsolation, (_void) -> fn.apply(tm, t));
+    }
+
+    // TODO: javadoc
+    default <T> void runInTransaction(TransactionIsolation transactionIsolation,
+                                      TransactionManagerAwareConsumerCapableOfThrowingException<T> fn, T t) throws Exception {
+        final TransactionManager tm = this;
+        runInTransactionInternal(transactionIsolation, (_void) -> {
+            fn.accept(tm, t);
+            return null;
+        });
+    }
+
+    default <R> R runInTransactionInternal(TransactionIsolation transactionIsolation,
+                                           FunctionCapableOfThrowingException<Void, R> fn) throws Exception {
+        boolean committed = false;
+        try {
+            beginTransaction(transactionIsolation);
+            R r = fn.apply(null);
+            commitTransaction();
+            committed = true;
+            return r;
+        } catch (IgnoreTransactionRollbackException e) {
+            commitTransaction();
+            committed = true;
+            throw convertThrowableToException(e.getCause());
+        } finally {
+            if (!committed) {
+                rollbackTransaction();
+            }
+        }
+    }
+
+    default Exception convertThrowableToException(Throwable t) {
+        // just assuming that t is not a type of Error to make life simpler
+        if (t instanceof Error) {
+            throw (Error) t;
+        } else if (t instanceof Exception) {
+            return (Exception) t;
+        } else {
+            return new RuntimeException(t);
+        }
+    }
+
+    @FunctionalInterface
+    interface FunctionCapableOfThrowingException<T, R> {
```

Comment:
FunctionCapableOfThrowingException takes one input and produces one output, I have seen scenarios in SAM where transaction block will take more than one input, any idea how should we go about handling such cases?